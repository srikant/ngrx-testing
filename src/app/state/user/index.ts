import { createFeatureSelector, createSelector } from '@ngrx/store';
import { State } from '@state/user/user.reducer';
import * as fromUser from './user.reducer';

export const selectUserState = createFeatureSelector<State>('user');

export const selectUserIds = createSelector(
  selectUserState,
  fromUser.selectUserIds
);
export const selectUserEntities = createSelector(
  selectUserState,
  fromUser.selectUserEntities
);
export const selectAllUsers = createSelector(
  selectUserState,
  fromUser.selectAllUsers
);
export const selectUserTotal = createSelector(
  selectUserState,
  fromUser.selectUserTotal
);
export const selectSelectedUserId = createSelector(
  selectUserState,
  fromUser.getSelectedUserId
);
export const selectSelectedUser = createSelector(
  selectUserEntities,
  selectSelectedUserId,
  (entities, selectedUserId) => entities && entities[selectedUserId]
);
