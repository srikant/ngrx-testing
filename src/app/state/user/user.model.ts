import * as faker from 'faker/locale/en_US';

export interface User {
  id: number;
  firstName: string;
  lastName: string;
}

export const generateUser = (): User => {
  return {
    id: faker.random.number(),
    firstName: faker.name.firstName(),
    lastName: faker.name.lastName()
  };
};

export const generateUsers = (
  count = faker.random.number({ min: 1, max: 20 })
): User[] => {
  return Array.apply(null, Array(count)).map(() => generateUser());
};
