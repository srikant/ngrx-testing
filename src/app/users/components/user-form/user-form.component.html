<form [formGroup]="form">
  <input type="text" placeholder="First Name" formControlName="firstName">
  <input type="text" placeholder="Last Name" formControlName="lastName">
  <button (click)="onSave()">Save</button>
</form>