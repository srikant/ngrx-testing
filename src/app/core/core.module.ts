import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import {
  ModuleWithProviders,
  NgModule,
  Optional,
  SkipSelf
} from '@angular/core';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { RootDbService } from './db/root.db.service';
import { UserService } from './services/user.service';

@NgModule({
  imports: [
    CommonModule,
    HttpClientModule,
    HttpClientInMemoryWebApiModule.forRoot(RootDbService, {
      delay: 500
    })
  ],
  declarations: [],
  providers: [UserService]
})
export class CoreModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CoreModule
    };
  }

  constructor(
    @Optional()
    @SkipSelf()
    parentModule?: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only'
      );
    }
  }
}
