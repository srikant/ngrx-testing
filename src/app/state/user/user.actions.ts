import { Update } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { User } from './user.model';

export enum UserActionTypes {
  LoadUser = '[User] Load User',
  LoadUserSuccess = '[User] Load User Success',
  LoadUserFail = '[User] Load User Fail',
  LoadUsers = '[User] Load Users',
  LoadUsersSuccess = '[User] Load Users Success',
  LoadUsersFail = '[User] Load Users Fail',
  AddUser = '[User] Add User',
  AddUserSuccess = '[User] Add User Success',
  AddUserFail = '[User] Add User Fail',
  AddUsers = '[User] Add Users',
  AddUsersSuccess = '[User] Add Users Success',
  AddUsersFail = '[User] Add Users Fail',
  UpdateUser = '[User] Update User',
  UpdateUserSuccess = '[User] Update User Success',
  UpdateUserFail = '[User] Update User Fail',
  UpdateUsers = '[User] Update Users',
  UpdateUsersSuccess = '[User] Update Users Success',
  UpdateUsersFail = '[User] Update Users Fail',
  DeleteUser = '[User] Delete User',
  DeleteUserSuccess = '[User] Delete User Success',
  DeleteUserFail = '[User] Delete User Fail',
  DeleteUsers = '[User] Delete Users',
  ClearUsers = '[User] Clear Users',
  SelectUser = '[User] Select User'
}

export class LoadUser implements Action {
  readonly type = UserActionTypes.LoadUser;

  constructor(public payload: { id: number }) {}
}

export class LoadUserSuccess implements Action {
  readonly type = UserActionTypes.LoadUserSuccess;

  constructor(public payload: { user: User }) {}
}

export class LoadUserFail implements Action {
  readonly type = UserActionTypes.LoadUserFail;

  constructor(public payload: { error: Error }) {}
}

export class LoadUsers implements Action {
  readonly type = UserActionTypes.LoadUsers;
}

export class LoadUsersSuccess implements Action {
  readonly type = UserActionTypes.LoadUsersSuccess;

  constructor(public payload: { users: User[] }) {}
}

export class LoadUsersFail implements Action {
  readonly type = UserActionTypes.LoadUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class AddUser implements Action {
  readonly type = UserActionTypes.AddUser;

  constructor(public payload: { user: Partial<User> }) {}
}

export class AddUserSuccess implements Action {
  readonly type = UserActionTypes.AddUserSuccess;

  constructor(public payload: { user: User }) {}
}

export class AddUserFail implements Action {
  readonly type = UserActionTypes.AddUserFail;

  constructor(public payload: { error: Error }) {}
}

export class AddUsers implements Action {
  readonly type = UserActionTypes.AddUsers;

  constructor(public payload: { users: Partial<User>[] }) {}
}

export class AddUsersSuccess implements Action {
  readonly type = UserActionTypes.AddUsersSuccess;

  constructor(public payload: { users: User[] }) {}
}

export class AddUsersFail implements Action {
  readonly type = UserActionTypes.AddUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class UpdateUser implements Action {
  readonly type = UserActionTypes.UpdateUser;

  constructor(public payload: { user: Partial<User> }) {}
}

export class UpdateUserSuccess implements Action {
  readonly type = UserActionTypes.UpdateUserSuccess;

  constructor(public payload: { update: Update<User> }) {}
}

export class UpdateUserFail implements Action {
  readonly type = UserActionTypes.UpdateUserFail;

  constructor(public payload: { error: Error }) {}
}

export class UpdateUsers implements Action {
  readonly type = UserActionTypes.UpdateUsers;

  constructor(public payload: { users: Partial<User>[] }) {}
}

export class UpdateUsersSuccess implements Action {
  readonly type = UserActionTypes.UpdateUsersSuccess;

  constructor(public payload: { update: Update<User>[] }) {}
}

export class UpdateUsersFail implements Action {
  readonly type = UserActionTypes.UpdateUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class DeleteUser implements Action {
  readonly type = UserActionTypes.DeleteUser;

  constructor(public payload: { id: string }) {}
}

export class DeleteUserSuccess implements Action {
  readonly type = UserActionTypes.DeleteUserSuccess;

  constructor(public payload: { id: string }) {}
}

export class DeleteUserFail implements Action {
  readonly type = UserActionTypes.DeleteUserFail;

  constructor(public payload: { error: Error }) {}
}

export class DeleteUsers implements Action {
  readonly type = UserActionTypes.DeleteUsers;

  constructor(public payload: { ids: string[] }) {}
}

export class ClearUsers implements Action {
  readonly type = UserActionTypes.ClearUsers;
}

export class SelectUser implements Action {
  readonly type = UserActionTypes.SelectUser;

  constructor(public payload: { id: number }) {}
}

export type UserActions =
  | LoadUser
  | LoadUserSuccess
  | LoadUserFail
  | LoadUsers
  | LoadUsersSuccess
  | LoadUsersFail
  | AddUser
  | AddUserSuccess
  | AddUserFail
  | AddUsers
  | AddUsersSuccess
  | AddUsersFail
  | UpdateUser
  | UpdateUserSuccess
  | UpdateUserFail
  | UpdateUsers
  | UpdateUsersSuccess
  | UpdateUsersFail
  | DeleteUser
  | DeleteUserSuccess
  | DeleteUserFail
  | DeleteUsers
  | ClearUsers
  | SelectUser;
