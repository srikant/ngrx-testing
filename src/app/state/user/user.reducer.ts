import { EntityAdapter, EntityState, createEntityAdapter } from '@ngrx/entity';
import { UserActionTypes, UserActions } from './user.actions';
import { User } from './user.model';

export interface State extends EntityState<User> {
  error?: Error;
  loading: boolean;
  selectedUserId: number;
}

export const adapter: EntityAdapter<User> = createEntityAdapter<User>();

export const initialState: State = adapter.getInitialState({
  loading: false,
  selectedUserId: null
});

export function reducer(state = initialState, action: UserActions): State {
  switch (action.type) {
    case UserActionTypes.AddUser:
    case UserActionTypes.AddUsers:
    case UserActionTypes.LoadUser:
    case UserActionTypes.LoadUsers:
    case UserActionTypes.UpdateUser:
    case UserActionTypes.UpdateUsers: {
      return { ...state, loading: true, error: undefined };
    }

    case UserActionTypes.AddUserSuccess:
    case UserActionTypes.LoadUserSuccess: {
      return adapter.addOne(action.payload.user, { ...state, loading: false });
    }

    case UserActionTypes.AddUserFail:
    case UserActionTypes.AddUsersFail:
    case UserActionTypes.LoadUserFail:
    case UserActionTypes.LoadUsersFail:
    case UserActionTypes.UpdateUserFail:
    case UserActionTypes.UpdateUsersFail: {
      return { ...state, error: action.payload.error, loading: false };
    }

    case UserActionTypes.AddUsersSuccess: {
      return adapter.addMany(action.payload.users, {
        ...state,
        loading: false
      });
    }

    case UserActionTypes.UpdateUserSuccess: {
      return adapter.updateOne(action.payload.update, {
        ...state,
        loading: false
      });
    }

    case UserActionTypes.UpdateUsersSuccess: {
      return adapter.updateMany(action.payload.update, {
        ...state,
        loading: false
      });
    }

    case UserActionTypes.DeleteUser: {
      return adapter.removeOne(action.payload.id, state);
    }

    case UserActionTypes.DeleteUsers: {
      return adapter.removeMany(action.payload.ids, state);
    }

    case UserActionTypes.LoadUsersSuccess: {
      return adapter.addAll(action.payload.users, { ...state, loading: false });
    }

    case UserActionTypes.ClearUsers: {
      return adapter.removeAll(state);
    }

    case UserActionTypes.SelectUser: {
      return { ...state, selectedUserId: action.payload.id };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedUserId = (state: State) => state.selectedUserId;

export const {
  selectIds: selectUserIds,
  selectEntities: selectUserEntities,
  selectAll: selectAllUsers,
  selectTotal: selectUserTotal
} = adapter.getSelectors();
