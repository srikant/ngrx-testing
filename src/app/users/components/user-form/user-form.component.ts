import {
  Component,
  EventEmitter,
  Input,
  OnChanges,
  Output,
  SimpleChanges
} from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { User } from '@state/user/user.model';

@Component({
  selector: 'app-user-form',
  templateUrl: './user-form.component.html',
  styleUrls: ['./user-form.component.scss']
})
export class UserFormComponent implements OnChanges {
  form: FormGroup;

  @Input() user: User;
  @Output() userChange = new EventEmitter<User>();

  constructor(private formBuilder: FormBuilder) {
    this.buildForm();
  }

  ngOnChanges(simpleChanges: SimpleChanges) {
    if (simpleChanges.user && simpleChanges.user.currentValue) {
      this.form.patchValue(simpleChanges.user.currentValue);
    }
  }

  onSave() {
    // verify form is valid
    if (this.form.invalid) {
      return;
    }

    this.userChange.emit({
      ...this.user,
      ...this.form.value
    });
  }

  private buildForm() {
    this.form = this.formBuilder.group({
      firstName: ['', Validators.required],
      lastName: ['', Validators.required]
    });
  }
}
