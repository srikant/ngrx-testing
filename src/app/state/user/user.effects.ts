import { Injectable } from '@angular/core';
import { UserService } from '@core/services/user.service';
import { Actions, Effect } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable, of } from 'rxjs';
import { catchError, exhaustMap, map } from 'rxjs/operators';
import {
  AddUser,
  AddUserFail,
  AddUserSuccess,
  LoadUser,
  LoadUserFail,
  LoadUserSuccess,
  LoadUsers,
  LoadUsersFail,
  LoadUsersSuccess,
  UpdateUser,
  UpdateUserFail,
  UpdateUserSuccess,
  UserActionTypes
} from './user.actions';

@Injectable()
export class UserEffects {
  @Effect()
  addUser: Observable<Action> = this.actions$
    .ofType<AddUser>(UserActionTypes.AddUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.addUser(payload.user)),
      map(user => new AddUserSuccess({ user })),
      catchError(error => of(new AddUserFail({ error })))
    );

  @Effect()
  loadUser: Observable<Action> = this.actions$
    .ofType<LoadUser>(UserActionTypes.LoadUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.getUser(payload.id)),
      map(user => new LoadUserSuccess({ user })),
      catchError(error => of(new LoadUserFail({ error })))
    );

  @Effect()
  loadUsers: Observable<Action> = this.actions$
    .ofType<LoadUsers>(UserActionTypes.LoadUsers)
    .pipe(
      exhaustMap(() => this.userService.getUsers()),
      map(users => new LoadUsersSuccess({ users })),
      catchError(error => of(new LoadUsersFail({ error })))
    );

  @Effect()
  updateUser: Observable<Action> = this.actions$
    .ofType<UpdateUser>(UserActionTypes.UpdateUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.updateUser(payload.user)),
      map(
        user =>
          new UpdateUserSuccess({
            update: {
              id: user.id,
              changes: user
            }
          })
      ),
      catchError(error => of(new UpdateUserFail({ error })))
    );

  constructor(private actions$: Actions, private userService: UserService) {}
}
