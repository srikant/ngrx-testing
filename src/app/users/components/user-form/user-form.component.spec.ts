import { SimpleChange } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { User, generateUser } from '@state/user/user.model';
import { UserFormComponent } from './user-form.component';

function newEvent(eventName: string, bubbles = false, cancelable = false) {
  const evt = document.createEvent('CustomEvent');
  evt.initCustomEvent(eventName, bubbles, cancelable, null);
  return evt;
}

describe('UserFormComponent', () => {
  let component: UserFormComponent;
  let fixture: ComponentFixture<UserFormComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [UserFormComponent],
      imports: [FormsModule, ReactiveFormsModule]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserFormComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should patch values into the form', () => {
    const user = generateUser();

    component.ngOnChanges({
      user: new SimpleChange(null, user, true)
    });

    expect(component.form.value).toEqual({
      firstName: user.firstName,
      lastName: user.lastName
    });
  });

  it('should emit the userChange event when submitted', () => {
    const user = generateUser();
    const firstName = 'Brian';
    const firstNameDebugEl = fixture.debugElement.query(
      By.css('input[formControlName="firstName"]')
    );
    const firstNameEl = firstNameDebugEl.nativeElement as HTMLInputElement;
    const buttonDebugEl = fixture.debugElement.query(By.css('button'));

    fixture.detectChanges();

    let updatedUser: User;
    component.userChange.subscribe(u => (updatedUser = u));

    component.user = user;
    component.ngOnChanges({
      user: new SimpleChange(null, user, true)
    });

    firstNameEl.value = firstName;
    firstNameEl.dispatchEvent(newEvent('input'));

    buttonDebugEl.triggerEventHandler('click', null);

    expect(updatedUser).toEqual({
      ...user,
      firstName
    });
  });
});
