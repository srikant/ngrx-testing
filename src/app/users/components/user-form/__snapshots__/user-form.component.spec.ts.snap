// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFormComponent should match snapshot 1`] = `
<app-user-form
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  userChange={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="firstName"
      ng-reflect-name="firstName"
      placeholder="First Name"
      type="text"
    />
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="lastName"
      ng-reflect-name="lastName"
      placeholder="Last Name"
      type="text"
    />
    <button>
      Save
    </button>
  </form>
</app-user-form>
`;
